### PR TITLE
Support opening local store with database on read-only filesystem

### PR DESCRIPTION
**Pull request by benradf**
# Motivation
Currently it is not possible to open a local store when its database is on a read-only filesystem. Obviously a store on a read-only filesystem cannot be modified, but it would still be useful to be able to query it.

# Context
This change adds a new `read-only` setting to `LocalStore`. When set to `true`, Nix will skip operations that fail when the database is on a read-only filesystem (acquiring `big-lock`, schema migration, etc), and the store database will be opened in _immutable_ mode.

> `immutable`: The immutable parameter is a boolean query parameter that indicates that the database file is stored on read-only media. When immutable is set, SQLite assumes that the database file cannot be changed, even by a process with higher privilege, and so the database is opened read-only and all locking and change detection is disabled. Caution: Setting the immutable property on a database file that does in fact change can result in incorrect query results and/or `SQLITE_CORRUPT` errors.

https://www.sqlite.org/c3ref/open.html

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Review by thufschmitt — CHANGES_REQUESTED** [review]
That's pretty nice. I expect things to break badly in a number of instances (anything that requires write access to the database), but it's still fairly useful.

I left some comments for the few things that stood out. Most importantly, I think this would deserve some tests (to at least ensure that some basic operations work properly)

**Comment by Ericson2314** [comment]
Agreed on tests, but also I told @benradf I'd be happy to write them per our general testing vs implementation breakdown. So not an oversight! :)

File: src/libstore/local-store.hh
Comment (Ericson2314) [reply src/libstore/local-store.hh:62]: I wonder if it's possible to have a read only database in immutable mode (flag already set) so anything that tries to open it r-w will fail?

Maybe I makes sense to ask upstream SQLite about this?

In the meantime, I think this is fine enough, or we can put it behind an unstable feature just to be sure.

Comment (benradf) [reply src/libstore/local-store.hh:62]: > .. have a read only database in immutable mode (flag already set) so anything that tries to open it r-w will fail?

If the database file itself could be marked as immutable somehow, that would be ideal. I'll look into it.

**Comment by edolstra** [comment]
Can you say a bit more about use cases for this?

Wouldn't this fail if the WAL journal exists?

**Comment by Ericson2314** [comment]
The use-case is a shared mount which is used a read only source of stuff. It could be a substitutor, or (spoiler!) lower layer of a layered store.

File: src/libstore/local-store.hh
Comment (roberth) [reply src/libstore/local-store.hh:62]: The SQLite docs state that `immutable` assumes that the file will not be written to, which is problematic as cachix 1.4 has found out the hard way.
SQLite does not have a read-only mode that can open the file in read only mode without this assumption; not when I looked for it, not when cachix did. I don't think they've asked SQLite upstream about it, but I'll check with them.

File: src/libstore/local-store.cc
Comment (roberth) [reply src/libstore/local-store.cc:310]: You could perhaps copy to a tempfile and migrate it in place there. That does destroy any chance of reading updates that were made after opening, but that doesn't seem to be @benradf's goal, and is unlikely to work well with `immutable` as well.

**Comment by roberth** [comment]
I think it'd be highly preferable to get metadata from a nix-daemon, especially based on cachix' experience with this approach. Their use case may have been more dynamic (`cachix watch-store`) but still this approach seems rather fragile.

**Comment by Ericson2314** [comment]
@roberth I'd be interested in understanding cachix's experience better, but yes the use-case here is not some sort of pub-sub on a growing or changing local store, but an actually immutable one "pre baked" and the mounted in a gazillion locations from cold storage.

This "bake and mount everywhere" approach I've seen more than once with large organizational users of Nix, so I think it is a decent one to add some support for, though I wouldn't mind marking it experiment due to these gotchas.

Ultimately I do agree it would be nice to have something more flexible too. Indeed, I was working on the IPFS stuff because our data model really should allow us to side step all the synchronization SQLite, NFS, etc. need to do to support arbitrary mutations. But the FUSE performance situation is...complicated, and in general fancier and novel things come with more uncertainty, so I think it's good to have some simple tools in our toolbox like this. Walking before running, etc.

**Comment by Ericson2314** [comment]
Oh and then as to the actual question, for a gazillion consumers of the pre-baked imagine, you don't really want to involve a a single `nix-daemon` as choke point. Having two communication channels not one is an annoying overhead system administrative, (especially when the second one is ssh!), and the single nix-daemon will be a very unoptimized choke point too (we don't have anything like a daemon read replica yet...and that would probably require something like this anyways :)).

File: src/libstore/local-store.hh
Comment (roberth) [reply src/libstore/local-store.hh]: They haven't asked upstream.

**Comment by domenkozar** [comment]
I've tried exactly this approach in Cachix, an attempt of not needing to link to libstore, but I couldn't get it to work.

With immutable=1 you're getting the snapshot of sqlite at the time of opening the database and in my testing, it didn't read the entries from WAL files.

**Comment by domenkozar** [comment]
It depends what you're doing but for Cachix,`cachix watch-exec nix-build` didn't work.

It would detect a new path has been built, open the store and the store path wasn't there according to the query.

I didn't dive into it very deep, but I assume it was part of WAL but not in the database itself.

We might take this upstream to sqlite and ask why does immutable not respect the WAL.

**Comment by fricklerhandwerk** [comment]
I'll take a look at docs and merge when they're good.

File: tests/read-only-store.sh
Comment (fricklerhandwerk) [inline tests/read-only-store.sh:4]:
```suggestion

needLocalStore "cannot open store read-only when daemon has already opened it writeable"

```

According to @ericson2314 this should (hopefully) fix the test

File: src/libstore/local-store.cc
Comment (fricklerhandwerk) [inline src/libstore/local-store.cc:294]: The header for `getSchema` could use a doc comment saying that `0` means no database.

File: src/libstore/local-store.cc
Comment (fricklerhandwerk) [inline src/libstore/local-store.cc:366]:
```suggestion
            throw Error("need to migrate to content-addressed schema, but this cannot be done in read-only mode");
```

Do not use abbreviations, this would be recursive jargon and very hard to parse for non-experts.

File: src/libstore/local-store.hh
Comment (fricklerhandwerk) [inline src/libstore/local-store.hh:60]:
```suggestion
          [`immutable` parameter](https://www.sqlite.org/c3ref/open.html) set. Do not use this unless the filesystem is read-only.
```
We may want to make unmistakably clear what is meant here.

File: src/libstore/local-store.hh
Comment (fricklerhandwerk) [inline src/libstore/local-store.hh:61]: Make this a warning admonition

```
> **Warning**
>
> ...
```

File: src/libutil/experimental-features.cc
Comment (fricklerhandwerk) [inline src/libutil/experimental-features.cc:221]: Would be great if you could also take on https://github.com/NixOS/nix/issues/8496 so we can link to parameters properly. That way the new feature will be way more useful.

File: src/libutil/experimental-features.cc
Comment (fricklerhandwerk) [inline src/libutil/experimental-features.cc:227]: Why is this even? Would be great if we could capture this qualitatively, for example (made up):

```suggestion
            Enabling this setting disables the locking required for safe concurrent
            access to greatly increase performance, so you should be certain that the database will not be changed.
```

**Review by fricklerhandwerk — APPROVED** [review]
Please fix the broken test, otherwise good to go from my side apart from minor suggestions you can apply directly. I strongly recommend using [one line per sentence ](https://nix.dev/contributing/documentation#markdown) as it's easier to review and make direct suggestions without having to rewrap everything.

File: src/libstore/local-store.hh
Comment (fricklerhandwerk) [inline src/libstore/local-store.hh:56]: Minor confusion for me: "This causes it to fail" could as well mean "Enabling this option causes it to fail", which you probably don't mean.
```suggestion
          Normally Nix will attempt to open the store database in read-write mode, even for querying (when write access is not needed), causing it to fail if the database is on a read-only filesystem.
```

File: src/libutil/experimental-features.cc
Comment (fricklerhandwerk) [reply src/libutil/experimental-features.cc:228]:
```suggestion
            Allow the use of the `read-only` parameter in local [store URLs](@docroot@/command-ref/new-cli/nix3-help-stores.md).
```

Not sure if the link is correct, but at least we have something to point people to. Implementing #8496 will break that link so it can be updated to point to the right thing.

File: src/libutil/experimental-features.cc
Comment (fricklerhandwerk) [inline src/libutil/experimental-features.cc:236]:
```suggestion
```

None of this needs to be documented in the experimental feature flag, but should instead be local to the actual setting. Also currently it's not only slightly confusing, because those sentences say almost (exactly?) the same thing but in very different words. Which version is authoritative?

**Review by fricklerhandwerk — APPROVED** [review]
Thanks for the updates. Last touch now, will merge after that.

File: src/libutil/experimental-features.cc
Comment (benradf) [reply src/libutil/experimental-features.cc:236]: The setting documentation was closer to authoritative so I've gone with that version (but with an extra sentence from the flag documentation that I think is useful).

File: src/libstore/sqlite.hh
Comment (Ericson2314) [inline src/libstore/sqlite.hh:32]: I think trailing comma is accepted here? 
```suggestion
    Immutable,
```

**Pull request merged**
